docs/api-docs for Redis (#17029)
hashicorp/vault: website/content/docs/secrets/databases/redis.mdx

---
layout: docs
page_title: Redis - Database - Secrets Engines
description: |-

> Redis is one of the supported plugins for the database secrets engine.
> This plugin generates database credentials dynamically based on configured
> roles for the Redis database.

Reviewer: This needs two spaces or it won't render correctly. Suggestion below should work.

```suggestion
  Redis is one of the supported plugins for the database secrets engine.
  This plugin generates database credentials dynamically based on configured
  roles for the Redis database.
```

---

# Redis Database Secrets Engine

Redis is one of the supported plugins for the database secrets engine. This
plugin generates database credentials dynamically based on configured roles for
the Redis database.

See the [database secrets engine](/docs/secrets/databases) docs for
more information about setting up the database secrets engine.

## Capabilities

| Plugin Name                 | Root Credential Rotation | Dynamic Roles | Static Roles | Username Customization |
| --------------------------- | ------------------------ | ------------- | ------------ | ---------------------- |
| `redis-database-plugin`     | Yes                      | Yes           | Yes          | No                     |

## Setup

1.  Enable the database secrets engine if it is not already enabled:

> ```bash

Reviewer 2: Two small comments:

The numbering for the setup list can actually be all `1.` instead of 1. 2. 3. etc. It renders the sequentially numbers automatically. What's nice about using this is that in the future if more steps are added, the author doesn't have to go back and reorder a bunch of numbers. The list under Dynamic roles and Static roles look to already be using this, if you want to look at how it renders as an example.

The language tag to use for shell code blocks is:
~~~
```shell-session
   ...
```
~~~
according to the README in the website directory of Vault. (Not sure the reason why if I'm honest with you, but there probably is one)

---

# Redis Database Secrets Engine

Redis is one of the supported plugins for the database secrets engine. This
plugin generates database credentials dynamically based on configured roles for
the Redis database.

See the [database secrets engine](/docs/secrets/databases) docs for
more information about setting up the database secrets engine.

## Capabilities

| Plugin Name                 | Root Credential Rotation | Dynamic Roles | Static Roles | Username Customization |
| --------------------------- | ------------------------ | ------------- | ------------ | ---------------------- |
| `redis-database-plugin`     | Yes                      | Yes           | Yes          | No                     |

## Setup

1.  Enable the database secrets engine if it is not already enabled:

    ```shell-session
    $ vault secrets enable database
    Success! Enabled the database secrets engine at: database/
    ```

    By default, the secrets engine will enable at the name of the engine. To
    enable the secrets engine at a different path, use the `-path` argument.

> 1.  Configure Vault with the proper plugin and connection configuration:

Reviewer 3: Should this be incrementing between sections?

Reviewer 3: Hmm seems like this is like this in most places and only a handful like [MySQL](http://localhost:3000/docs/secrets/databases/mysql-maria) increment correctly. Could this be a general bug that was missed some time ago?

Seems odd to me that numerical bullet points all use 1.

    ```shell-session
    $ vault write database/config/my-redis-database \
      plugin_name="redis-database-plugin" \
      host="localhost" \
      port=6379 \
      username="vaultuser" \
      password="vaultpass" \
      allowed_roles="my-*-role"

>     ```

Reviewer 3: Should we re-use the same sample values from the api-doc page for consistency and to let users see the relation between both methods?

Author: 👍

1.  You should consider rotating the admin password. Note that if you do, the
    new password will never be made available through Vault, so you should
    create a Vault-specific database admin user for this.

    ```shell-session
    vault write -force database/rotate-root/my-redis-database
    ```

## Usage

After the secrets engine is configured, configure dynamic and static roles
to enable generating credentials.

### Dynamic roles

1.  Configure a dynamic role that maps a name in Vault to a JSON string
    containing the Redis ACL rules which are documented [here](https://redis.io/commands/acl-cat) or in the output
    of the `ACL CAT` redis command.

    ```shell-session
    $ vault write database/roles/my-dynamic-role \
        db_name="my-redis-database" \

>         creation_statements='["+@admin"]' \

Reviewer 3: Should we give an example that shows more options and good practices?

Something like '["~*", "-@admin", "+@read"]' to show how to pass multiple statements.

        default_ttl="5m" \
        max_ttl="1h"
    Success! Data written to: database/roles/my-dynamic-role
    ```

Note that if a creation_statement is not provided the user account will
default to a read only user, `'["~*", "+@read"]'` that can read any key.

> 1.  Generate a new credential by reading from the `/creds` endpoint with the name

Reviewer 3: Can you really say "generate a new credential" or it should be "generate new credentials"?

I almost never see the singular form but maybe this is valid, English is not my first language so maybe it's just me :sweat_smile:

Author: +1 on the first language 🤣 

All of the docs are phrased as "generate a new credential", so I think for the consistency we should leave as is?

| `redis-database-plugin`     | Yes                      | Yes           | Yes          | No                     |

## Setup

1.  Enable the database secrets engine if it is not already enabled:

    ```shell-session
    $ vault secrets enable database
    Success! Enabled the database secrets engine at: database/
    ```

    By default, the secrets engine will enable at the name of the engine. To
    enable the secrets engine at a different path, use the `-path` argument.

1.  Configure Vault with the proper plugin and connection configuration:

    ```shell-session
    $ vault write database/config/my-redis-database \
      plugin_name="redis-database-plugin" \
      host="localhost" \
      port=6379 \
      tls=true \
      ca_cert="$CACERT"
      username="user" \
      password="pass" \
      allowed_roles="my-*-role"
    ```

1.  You should consider rotating the admin password. Note that if you do, the
    new password will never be made available through Vault, so you should
    create a Vault-specific database admin user for this.

    ```shell-session
    vault write -force database/rotate-root/my-redis-database
    ```

## Usage

> After the secrets engine is configured, configure dynamic and static roles
> to enable generating credentials.

Reviewer 2: ```suggestion
After the secrets engine is configured, write dynamic and static roles
to Vault to enable generating credentials.
```
"Configure" back to back like that is a bit jarring to me.

### Dynamic roles

> 1.  Configure a dynamic role that maps a name in Vault to a JSON string
>     containing the Redis ACL rules which are documented [here](https://redis.io/commands/acl-cat) or in the output
>     of the `ACL CAT` redis command.

Reviewer 2: ```suggestion
1.  Configure a dynamic role that maps a name in Vault to a JSON string
    containing the Redis ACL rules, which are either documented [here](https://redis.io/commands/acl-cat) or in the output
    of the `ACL CAT` Redis command.
```

    ```shell-session
    $ vault write database/roles/my-dynamic-role \
        db_name="my-redis-database" \
        creation_statements='["+@admin"]' \
        default_ttl="5m" \
        max_ttl="1h"
    Success! Data written to: database/roles/my-dynamic-role
    ```

> Note that if a creation_statement is not provided the user account will
> default to a read only user, `'["~*", "+@read"]'` that can read any key.

Reviewer 2: ```suggestion
    Note that if a creation_statement is not provided the user account will
    default to a read only user, `'["~*", "+@read"]'` that can read any key.
```

This needs to be indented to not break the automatic numbering of the steps.

> 1.  Generate a new credential by reading from the `/creds` endpoint with the name

Reviewer 2: Maybe this should be more specific and say "generate a new set of credentials by..." since it makes both a username and password?

Author: Thank you for suggestions and reviewing! 
I think I will leave this one as is, because I just checked all of the other database secrets engine documents, and every single one of them is formulated like this. So, to stay consistent I think that we should do the same for this one? Would you agree? @robmonte

Reviewer 2: I think _ideally_ the others would also be updated to reflect this, but it's such a trivial detail that I don't think it matters much. And updating the wording of other pages is outside of the scope here.

FWIW, I don't think things always need to stay consistent when the change is an improvement, but for this minor wording difference I am happy to leave it this way if you prefer it to be. 👍

Author: I think that you are right, eventually we can update all of them. I went ahead and updated it to say "a new set of credentials by". Thank you so much!

    of the role:

    ```shell-session
    $ vault read database/creds/my-dynamic-role
    Key                Value
    ---                -----
    lease_id           database/creds/my-dynamic-role/OxCTXJcxQ2F4lReWPjbezSnA
    lease_duration     5m
    lease_renewable    true
    password           dACqHsav6-attdv1glGZ
    username           V_TOKEN_MY-DYNAMIC-ROLE_YASUQUF3GVVD0ZWTEMK4_1608481717
    ```

### Static roles

> 1.  Configure a static role that maps a name in Vault to an existing redis

Reviewer 2: ```suggestion
1.  Configure a static role that maps a name in Vault to an existing Redis
```
